File: 380.insert-delete-get-random-o-1.py
#
# @lc app=leetcode id=380 lang=python3
#
# [380] Insert Delete GetRandom O(1)
#

# Time:  O(1)
# Space: O(n)

import random


# @lc code=start
class RandomizedSet:
    def __init__(self):
        self.list = []
        self.dict = {}

    def insert(self, val: int) -> bool:
        if val in self.dict:
            return False
        self.list.append(val)
        self.dict[val] = len(self.list) - 1
        return True

    def remove(self, val: int) -> bool:
        if val not in self.dict:
            return False

        self.dict[self.list[-1]] = self.dict[val]

        temp = self.list[self.dict[val]]
        self.list[self.dict[val]] = self.list[-1]
        self.list[-1] = temp

        self.dict.pop(val)
        self.list.pop()
        return True

    def getRandom(self) -> int:
        return self.list[random.randint(0, len(self.list) - 1)]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
# @lc code=end

File: 380.insert-delete-get-random-o-1.js
/*
 * @lc app=leetcode id=380 lang=javascript
 *
 * [380] Insert Delete GetRandom O(1)
 */

// @lc code=start
/**
 * Initialize your data structure here.
 */
var RandomizedSet = function () {
  this.list = [];
  this.dict = {};
};

/**
 * Inserts a value to the set. Returns true if the set did not already contain the specified element.
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.insert = function (val) {
  if (val in this.dict) {
    return false;
  }
  this.list.push(val);
  this.dict[val] = this.list.length - 1;
  return true;
};

/**
 * Removes a value from the set. Returns true if the set contained the specified element.
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.remove = function (val) {
  if (!(val in this.dict)) {
    return false;
  }

  this.dict[this.list[this.list.length - 1]] = this.dict[val];

  const temp = this.list[this.dict[val]];
  this.list[this.dict[val]] = this.list[this.list.length - 1];
  this.list[this.list.length - 1] = temp;

  delete this.dict[val];
  this.list.pop();
  return true;
};

/**
 * Get a random element from the set.
 * @return {number}
 */
RandomizedSet.prototype.getRandom = function () {
  return this.list[Math.trunc(Math.random() * this.list.length)];
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */
// @lc code=end
